### DISCUSSION TITLE
Does renovate cache private dependencies?

### DISCUSSION THREAD

**Discussion by eriksw** [renovatebot/renovate#24386] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Wanted end result.

I would like a manually-triggered Renovate run to pick up new images that I've pushed to private (Google Artifact Registry, in this case) Docker repos, without unnecessary delay.

It appears that Renovate may be caching the results of version lookups from private sources. Can you confirm whether that is that the case? Is there anything I can tune?

In this particular case it's only a moderate annoyance—if I was desperate to get the new docker image into use ASAP I could update the pin myself after the image build completes.

However, we also have a production use cases where a long-lived cache by Renovate could add nontrivial unnecessary delay in rolling out important fixes from public images.

<details><summary>Lenghty use case where caching private dep lookups hurts us</summary>

We need to deploy to a location (China) that doesn't have reliable outbound access to the main Docker Hub (or anything else out of the country). We work around this by having a GitOps/CD pipeline that copies images from public Docker Registry into our private registry. The on-merge pipeline of that repo first copies the images into an in-China ECR registry, then _after_ the image has definitely been pushed to the China, it copies them into a private Artifact Registry. (This pipeline is driven by Renovate making updates to the manifest of what to copy.)

We then have Renovate update our "ops-control" ("what should be running everywhere _right now_") repo by detecting the updated images being available in the last (second) Docker registry that the images get copied to by the first pipeline and making PRs.

If renovate has a long cache of what's in the private Docker Artifact Registry, that means that otherwise-deployable updates will sit there with an unnecessary delay before Renovate creates the update to the "ops-control" repo to put them into use. If there's an 0day for a public-facing Envoy container or something like that, that's not ideal.
</details>

### What you tried so far.

I've tried manually requesting runs of the repo a couple times after pushing an update (changing what `:latest` points to in a Docker registry), but both of those runs don't result in Renovate noticing the new version and making an update.

See notes in the logs below. This is from Job ID `a7fd789f-1f70-46f4-9d8d-78602a902372` in the hosted app, sorry for the redactions.

The updated docker image was pushed to the repo (the `:latest` tag was changed) at `2023-09-12 20:04:43.684 UTC`. That Renovate job completed at `Sep 12, 2023, 13:22:57` browser time, duration `20s`, ~17 minutes after the new version was available. 

The followup Renovate job `f487566d-7f71-4e4e-ab9c-264e5226e0ba` which completed at `Sep 12, 2023, 13:54:14` (browser time, ~50 minutes after the new version was available) _did_ see the new version and correctly made a PR.

### Relevant debug logs

<details><summary>Logs</summary>

I have a hostRule where I have a service account key encrypted for use by Renovate. The following snippet is from `Debug: Post-massage config`

```
    "hostRules": [
      {
        "matchHost": "us-docker.pkg.dev",
        "authType": "Basic",
        "encrypted": {
          "token": "***********"
        }
      }
    ],
```

This config does appear to get picked up.

```
DEBUG: Adding token authentication for us-docker.pkg.dev to hostRules
```

The following are all the output bits that I think are relevant to the lookup/processing of the specific image that this run didn't see an updated version for:

```
DEBUG: Dependency us-docker.pkg.dev/MYPROJECT/MYREPO/MYIMAGE has unsupported/unversioned value latest (versioning=docker)
```

from `DEBUG: packageFiles with updates`:

```
    "docker-compose": [
      {
        "deps": [
          {
            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
            "currentDigest": "sha256:b801f7a2806bb03da85bcab23e9f6ccc73e1469f62bc2ede59be9604ae72a8e1",
            "currentValue": "latest",
            "datasource": "docker",
            "depName": "us-docker.pkg.dev/MYPROJECT/MYREPO/MYIMAGE",
            "packageName": "us-docker.pkg.dev/MYPROJECT/MYREPO/MYIMAGE",
            "replaceString": "us-docker.pkg.dev/MYPROJECT/MYREPO/MYIMAGE:latest@sha256:b801f7a2806bb03da85bcab23e9f6ccc73e1469f62bc2ede59be9604ae72a8e1",
            "versioning": "docker",
            "warnings": [],
            "updates": []
          },
```

</details>

**Comment by rarkins** [comment, marked as answer]
Renovate indeed caches lookup results and it's an essential requirement to be able to scale, avoid rate limits, etc. If you're going to make statements like:

> Renovate could add nontrivial unnecessary delay in rolling out important fixes from public images.

..and then use 17 minutes as an example, then I think that your expectations - particularly of a _free_ service - are unreasonable. I'm pretty sure there can be much higher delays at times due to caching.

In other words, what you write is "unnecessary" is fact _necessary_ to run a public service like this. If you can can change your opinion about that, then we can continue discussing details. If not, I suggest you run your own infrastructure where you can turn off all caching if you wish.
